src/dowel/csv_output.py

```diff
@@ -48,7 +51,25 @@ def record(self, data, prefix=''):
                            'Did you change key sets after your first '
                            'logger.log(TabularInput)?'.format(
                                set(self._fieldnames), set(to_csv.keys())))
-
+                with open(self._filename, "r") as old:
+                    old_data = csv.DictReader(old)
+                    
+                    # Update the header
+                    self._fieldnames = set(self._longestHeader)|set(to_csv.keys())
+                    self._longestHeader = self._fieldnames
+
+                    # Change the writer with the new keys
+                    self._writer = csv.DictWriter(
+                        self._log_file,
+                        fieldnames=self._fieldnames,
+                        extrasaction='ignore')
+
+                    # Write the old data from the start of the file
+                    self._log_file.seek(0)
+                    self._writer.writeheader()
+                    for i in old_data:
```

Comment:
`DictReader` reads data on-demand from the file here, as it's being overwritten. Is that correct? It's not obvious to me why that would work.

Reply:
_log_file.seek(0) will change the file pointer to the beginning of the file. Thus, it will rewrite the file from the beginning. Does it make sense? Thank you!